auth/src/main/java/com/firebase/ui/auth/util/AnonymousUpgradeUtils.java

```diff
@@ -0,0 +1,120 @@
+package com.firebase.ui.auth.util;
+
+import android.app.Activity;
+import android.support.annotation.NonNull;
+import android.support.annotation.RestrictTo;
+
+import com.firebase.ui.auth.data.model.FlowParameters;
+import com.google.android.gms.tasks.Continuation;
+import com.google.android.gms.tasks.Task;
+import com.google.firebase.FirebaseApp;
+import com.google.firebase.auth.AuthCredential;
+import com.google.firebase.auth.AuthResult;
+import com.google.firebase.auth.EmailAuthProvider;
+import com.google.firebase.auth.FirebaseAuth;
+import com.google.firebase.auth.FirebaseAuthUserCollisionException;
+
+/**
+ * Utilities to help with Anonymous user upgrade.
+ */
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+public class AnonymousUpgradeUtils {
+
+    /**
+     * Uses type system to enforce the proper failure listener.
+     */
+    public static class UpgradeTaskWrapper<T> {
```

Comment:
Yeah that was the plan, to make sure that we always handle that error case.  But you're right that it doesn't quite work, because in some places we have continuations.